server/src/main/java/org/opensearch/tasks/Task.java

```diff
@@ -195,6 +232,100 @@ public Status getStatus() {
+            if (threadResourceInfo.getStatsType() == statsType && threadResourceInfo.isActive()) {
+                throw new IllegalStateException(
+                    "unexpected active thread resource entry present [" + threadId + "]:[" + threadResourceInfo + "]"
+                );
+            }
+        }
+        threadResourceInfoList.add(new ThreadResourceInfo(statsType, resourceUsageMetrics));
+    }
+
+    /**
+     * This method is used to update the resource consumption stats so that the data isn't too stale for long-running task.
+     * If active thread entry is present in the list, the entry is updated. If one is not found, it throws an exception.
+     * @param threadId ID of the thread
+     * @param statsType stats type
+     * @param resourceUsageMetrics resource consumption metrics of the thread
+     * @throws IllegalStateException if no matching active thread entry was found.
+     */
+    public void updateThreadResourceStats(long threadId, ResourceStatsType statsType, ResourceUsageMetric... resourceUsageMetrics) {
+        final List<ThreadResourceInfo> threadResourceInfoList = resourceStats.get(threadId);
+        if (threadResourceInfoList != null) {
+            for (ThreadResourceInfo threadResourceInfo : threadResourceInfoList) {
+                // the active entry present in the list is updated
+                if (threadResourceInfo.getStatsType() == statsType && threadResourceInfo.isActive()) {
+                    threadResourceInfo.recordResourceUsageMetrics(resourceUsageMetrics);
+                    return;
+                }
+            }
+        }
+        throw new IllegalStateException("cannot update if active thread resource entry is not present");
+    }
+
+    /**
+     * Record the thread's final resource consumption values.
+     * If active thread entry is present in the list, the entry is updated. If one is not found, it throws an exception.
+     * @param threadId ID of the thread
+     * @param statsType stats type
+     * @param resourceUsageMetrics resource consumption metrics of the thread
+     * @throws IllegalStateException if no matching active thread entry was found.
+     */
+    public void stopThreadResourceTracking(long threadId, ResourceStatsType statsType, ResourceUsageMetric... resourceUsageMetrics) {
+        final List<ThreadResourceInfo> threadResourceInfoList = resourceStats.get(threadId);
+        if (threadResourceInfoList != null) {
+            for (ThreadResourceInfo threadResourceInfo : threadResourceInfoList) {
+                if (threadResourceInfo.getStatsType() == statsType && threadResourceInfo.isActive()) {
+                    threadResourceInfo.setActive(false);
+                    threadResourceInfo.recordResourceUsageMetrics(resourceUsageMetrics);
+                    return;
+                }
+            }
+        }
+        throw new IllegalStateException("cannot update final values if active thread resource entry is not present");
+    }
```

Comment:
Not sure setting a flag to inactive for disabling resource tracking is the optimal way here. Since consumers would still be trying to update resource consumption while only to know its throw-away.

Reply:
@Bukhtawar This for tracking thread level updates as multiple threads can execute the same task. I have a separate commit (follow up PR) which adds a setting to disable the feature in which we avoid the overhead of fetching the resource info from Thread MX bean itself.

Reply:
```isActive``` field in the ```ThreadResourceInfo``` indicates the current status of the thread.
Tushar's changes will have a setting to disable the tracking feature.

Reply:
When we `stopThreadResourceTracking ` we are marking every thread as inactive. I couldn't find anywhere else that we are setting it inactive.  Is the caller of the method expected to be calling it when a task finishes?

Reply:
This method is expected to be called when the thread finishes work for this task. There is a possibility that the same thread can be assigned to the task again, in that case a new entry would be added to the list.